### ISSUE TITLE
Less than 25% of first 100000 reads have cell barcodes from barcodes file

### ISSUE THREAD

**Issue by brianpenghe** [wheaton5/souporcell#112]
I was running souporcell using this command:
```
singularity exec -B $PWD ~/tools/souporcell/souporcell.sif souporcell_pipeline.py \
    -i ${SAMPLE}/Aligned.sortedByCoord.out.bam -b ${SAMPLE}/counts/Gene/cr3/barcodes.tsv \
    -f ~/refseq/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa -t 16 -o souporcell${SAMPLE} \
    --common_variants ~/tools/souporcell/common_variants_grch38.vcf --skip_remap True -k 2
```
But it complained this:
```
checking modules
imports done
checking bam for expected tags
Traceback (most recent call last):
  File "/opt/souporcell/souporcell_pipeline.py", line 106, in <module>
    assert float(num_cb_cb) / float(num_read_test) > 0.05, "Less than 25% of first 100000 reads have cell barcodes from barcodes file, is this the correct barcode file? turn on --ignore True to ignore"
AssertionError: Less than 25% of first 100000 reads have cell barcodes from barcodes file, is this the correct barcode file? turn on --ignore True to ignore
```
Since my sample has 50% of its reads in cells, there's no way that the first 25% only contain soup.

I tried --ignore True, but that led to a result with mostly unassigned cells.

Any ideas?

**Comment by wheaton5** [comment]
Probably some bam format issue. Does your bam contain CB tags? Can you just print off 10 reads from the bam to look at?

**Comment by brianpenghe** [comment]
```
SRR11402938.1.159978018 16      1       11997   1       76M     *       0       0       TCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTT    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:74 nM:i:0
SRR11402938.1.197561979 16      1       14356   1       24S52M  *       0       0       TTTTTTTTTTTTTTTTTTTTTTTTTTTTCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTC    FFFFFFFFFF,FFFFFFFFFFFFF::FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF,FFFFFFFFFFFF:,FF    NH:i:4  HI:i:1  AS:i:49 nM:i:1  CB:Z:CTCCTCTAACCA       UB:Z:GAGGAGTT
SRR11402938.1.19688104  16      1       14362   1       8S68M   *       0       0       TTTTTTTTTCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTATTGATTGG    FFFFFFFFFFFFFFF,FFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF:FFFF:FFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:66 nM:i:0  CB:Z:GTCGACCAAGTA       UB:Z:CGGGGTTT
SRR11402938.1.31167270  16      1       14362   1       15S61M  *       0       0       TTTTTTTTTTTTTTTTCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTAT    FFFFFFFFFFFFF::FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:60 nM:i:0  CB:Z:GACGGCATGCCT       UB:Z:CGGGACAA
SRR11402938.1.75779446  16      1       14362   1       25S51M  *       0       0       TTTTTTTTTTTTTTTTTTTTTTTTTTCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTC    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:50 nM:i:0  CB:Z:AATAACCCATAA       UB:Z:GTGCGTTT
SRR11402938.1.130330149 16      1       14362   1       8S68M   *       0       0       TTTTTTTTTCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTATTGATTGG    :FFF::F:FFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFF,FF,FFFFFFFF,FFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:66 nM:i:0  CB:Z:GTCGACCAAGTA       UB:Z:CGGGGTTT
SRR11402938.1.140541039 16      1       14362   1       25S51M  *       0       0       TTTTTTTTTTTTTTTTTTTTTTTTTTCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTC    F:FF:FFF:FFFFFFFFFF:F:FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFF:F:FFFFFF    NH:i:4  HI:i:1  AS:i:50 nM:i:0  CB:Z:AATAACCCATAA       UB:Z:GTGCGTTT
SRR11402938.1.158036615 16      1       14364   1       24S52M  *       0       0       TTTTTTTTTTTTTTTTTTTTTTTTCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGT    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:51 nM:i:0  CB:Z:ATAATCTTAATT       UB:Z:GAGATTTT
SRR11402938.1.27829082  16      1       14365   1       4S72M   *       0       0       TTTTTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTATTGATTGGTGTGCCG    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:70 nM:i:0  CB:Z:GTAAGCCAATCA       UB:Z:AGGGGTTT
SRR11402938.1.43736407  16      1       14365   1       18S58M  *       0       0       TTTTTTTTTTTTTTTTTTTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTAT    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:4  HI:i:1  AS:i:57 nM:i:0  CB:Z:CTGCTGAGGCAT       UB:Z:CAGGATAA

```

**Comment by wheaton5** [comment]
and are those barcodes in your barcodes.tsv file?

**Comment by brianpenghe** [comment]
OK indeed none of them are in my barcodes.tsv file. I'll have to figure out why from my end. Thanks!

**Comment by ktpolanski** [comment]
So as an aftermath of this, it turns out that STAR 2.7.3a created random gibberish `CB:` tags when running DropSeq data with `--soloCBwhitelist None`, yet somehow had the correct barcodes in the count matrices. So I took the filtered cell space and used that as the whitelist, re-ran STAR, and got a BAM with correct `CB:` tags out of it.

I then threw it at souporcell, which complained. Differently than for @brianpenghe:
```
AssertionError: Less than 50% of first 100000 reads have cell barcode tag (CB), turn on --ignore True to ignore
```
I checked by hand, and out of the first 100,000 reads, 46,551 had a `CB:`. This seemed acceptable to me, as the barcode whitelist was limited to just the filtered cells. As such, I added `--ignore True`, and got an output that looks like a model case of there not being two genotypes - a three-way split between 0/1/unassigned.

So after all this introduction comes the reason for my posting - would having this sort of truncated barcode whitelist jeopardise souporcell output in any way? My gut feeling is that as long as I have `CB:` entries for the barcodes from `barcodes.tsv` it should be able to find the genotypes for them, right?
